okhttp/src/test/java/io/grpc/transport/okhttp/OkHttpClientTransportTest.java

```diff
@@ -450,6 +442,166 @@ public void streamIdExhausted() throws Exception {
+    new Thread(new Runnable() {
+      @Override
+      public void run() {
+        clientTransport.newStream(method, new Metadata.Headers(), listener);
+        newStreamReturn.countDown();
+      }
+    }).start();
+    assertStreamPending(1);
+
+    clientTransport.shutdown();
+
+    assertTrue("newStream() call is still blocking",
+        newStreamReturn.await(TIME_OUT_MS, TimeUnit.MILLISECONDS));
+    listener.waitUntilStreamClosed();
+    assertEquals(Status.INTERNAL.getCode(), listener.status.getCode());
+    assertEquals(0, clientTransport.getPendingStreamSize());
+  }
+
+  @Test
+  public void pendingStreamFailedByIdExhausted() throws Exception {
+    int startId = Integer.MAX_VALUE - 4;
+    clientTransport = new OkHttpClientTransport(
+        executor, frameReader, frameWriter, startId, new MockSocket(frameReader));
+    clientTransport.start(listener);
+    frameHandler = clientTransport.getHandler();
+    streams = clientTransport.getStreams();
+    setMaxConcurrentStreams(1);
+
+    final MockStreamListener listener1 = new MockStreamListener();
+    final MockStreamListener listener2 = new MockStreamListener();
+    final MockStreamListener listener3 = new MockStreamListener();
+
+    OkHttpClientStream stream1 =
+        clientTransport.newStream(method, new Metadata.Headers(), listener1);
+
+    final CountDownLatch newStreamReturn2 = new CountDownLatch(1);
+    final CountDownLatch newStreamReturn3 = new CountDownLatch(1);
+    // The second and third stream should be pending.
+    new Thread(new Runnable() {
+      @Override
+      public void run() {
+        clientTransport.newStream(method, new Metadata.Headers(), listener2);
+        newStreamReturn2.countDown();
+      }
+    }).start();
+    new Thread(new Runnable() {
+      @Override
+      public void run() {
+        clientTransport.newStream(method, new Metadata.Headers(), listener3);
+        newStreamReturn3.countDown();
+      }
+    }).start();
+
+    assertStreamPending(2);
+    assertEquals(1, streams.size());
+    assertEquals(startId, (int) stream1.id());
+
+    // Now finish stream1, stream2 should be started and exhaust the id,
+    // so stream3 should be failed.
+    stream1.cancel();
+    assertTrue("newStream() call for stream2 is still blocking",
+        newStreamReturn2.await(TIME_OUT_MS, TimeUnit.MILLISECONDS));
+    assertTrue("newStream() call for stream3 is still blocking",
+        newStreamReturn3.await(TIME_OUT_MS, TimeUnit.MILLISECONDS));
+    listener1.waitUntilStreamClosed();
+    listener3.waitUntilStreamClosed();
+    assertEquals(Status.INTERNAL.getCode(), listener3.status.getCode());
+    assertEquals(0, clientTransport.getPendingStreamSize());
+    assertEquals(1, streams.size());
+    OkHttpClientStream stream2 = streams.get(startId + 2); 
+    assertNotNull(stream2);
+    stream2.cancel();
+  }
+
+  private void assertStreamPending(int expected) throws Exception {
```

Comment:
Could we rename this to waitForStreamPending() or similar? assert isn't obvious that it blocks.